An example ECN validation algorithm (#3320)
quicwg/base-drafts: draft-ietf-quic-transport.md

this happens, and if validation succeeds, the local reference counts MUST be
increased to match the counts in the ACK frame.

> 

Reviewer: Related note: I just realised that we don't test the case where ALL ECT get remarked to CE which should also lead to failure.

later time in the connection.

> #### Exemplary ECN Validation Algorithm {#ecn-alg}

Reviewer: If we want an algorithm like this I think this should go into the appendix.

Each time that an endpoint commences sending on a new network path, it
determines whether the path supports ECN.  If the path supports ECN, the goal is
to use ECN.  Endpoints might also periodically reassess a path that was
determined not support ECN.

This section describes one method for testing new paths.  This algorithm is
intended to show how a path might be tested for ECN support.  Endpoints can
implement different methods.

The path has an ECN state that is one of "testing", "unknown", "failed", or
"capable".  In the "testing" and "capable" states the endpoint sends packets

> with an ECT(0) marking; otherwise, the endpoint sends unmarked packets.

Reviewer: I would say: "the endpoints sends packet with an ECT marking, by default ECT(0);"

To start testing a path, the ECN state is set to "testing" and existing ECN
counts are remembered as a baseline.  It should not be necessary to disregard
the effect of packets sent prior to starting testing, though it is necessary for
a sender to remember what markings were used for every packet that is
acknowledged; see {{ecn-ack}}.

The testing period runs for a number of packets or round trip times as
determined by the endpoint.  During this time, packets sent are marked with
ECT(0).  The goal is to limit the duration of the testing period, but to ensure

> that enough marked packets are sent that it is likely that ECN counts will

Reviewer 2: ```suggestion
that enough marked packets are sent for received ECN counts to
```

provide a clear indication of how the path treats marked packets.

<!-- Do we need a more concrete recommendation here?  For instance, I might say
"Endpoints could test with packets that amount to between 1 to 2 times the

> initial congestion window over a period between 1 to 2 times the estimated RTT."

Reviewer 2: This is an example, and as such, it's useful for us to provide a specific value here. We should also expect that anyone implementing this is likely to use these exemplary values.

> -->

Reviewer: Currently the ECN validation section says "To reduce the chances of misinterpreting congestive loss as packets
   dropped by a faulty network element, an endpoint could set the ECT(0)
   codepoint in the first ten outgoing packets on a path, or for a
   period of three RTTs, whichever occurs first."
Why is that not sufficient?

> After the testing period ends, the ECN state for the path becomes "unknown".

Reviewer 2: Why does this have to go to "unknown" and not directly to "capable" or "failed"?

Author: Though Mirja will disagree, this algorithm is extremely conservative, consistent with the existing test, and it disables ECT markings after this period in case there is a black hole.  I suspect that some implementations will just always mark and that's cool, but that's somewhat simpler, I think.  In that case, the algorithm is "mark unless validation on that path has recently failed".

From the "unknown" state, successful validation of the ECN counts an ACK frame
(see {{ecn-ack}}) causes the ECN state for the path to become "capable", unless

> no marked packet has been acknowledged.

Author: This extra condition caught me by surprise.  The problem is that if every packet you mark is lost, the ECN counts remain valid, but you have a fair indication that the path is a black hole.  That's not worth belabouring because black holes are exceedingly rare, but the extra condition seems necessary even if we don't provide a whole lot of explanation.

Reviewer: Given that black-holing is rare, I'd rather keep simply all packet ECT-marked until a failure is detect. If black-holing is happening, that's actually the easiest case to detect (as you will have a time-out). Moreover if you plan to send more than one cwnd of packets with ECT marks in testing, you will get blocked anyway as you can't send anymore packets without receiving an ack.

Reviewer 2: You will be blocked, but only until the PTO timer fires. There is a question still of what to do afterwards. While it is rare, I suspect we need this fallback as a failsafe.

Reviewer: My thinking was that if you get an PTO in testing to go straight to failed and don't sent any ECT markings anymore.

Author: I don't think that works without additional constraints or checks as you could exit the testing state before any PTO might occur.

Reviewer: I think the idea would be to only have three states: "Unknown/testing" and "capable" where ETC marking is enabled, "failed" where ECT marking is disabled and you stay in "unknown/testing" until the path has be validate for failed.

> At any point, if validation of ECN counts fails, the ECN state becomes "failed".

Reviewer 2: ```suggestion
If validation of ECN counts fails at any time, the endpoint records ECN state for the path as "failed".
```

   return candidate_pn
~~~

# Sample ECN Validation Algorithm {#ecn-alg}

Each time an endpoint commences sending on a new network path, it determines
whether the path supports ECN; see {{ecn}}.  If the path supports ECN, the goal
is to use ECN.  Endpoints might also periodically reassess a path that was

> determined not support ECN.

Reviewer 3: ```suggestion
determined to not support ECN.
```
